src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php

```diff
@@ -755,6 +755,11 @@ public function RecalculateBasket()
         $this->RecalculateVAT();
         $this->dCostTotal = $this->dCostArticlesTotalAfterDiscounts + $this->dCostShipping + $this->dCostPaymentMethodSurcharge;
         $this->RecalculateVouchers();
+        // Since vouchers are recalculated twice, we need to remember all calls to RemoveInvalidVouchers internally to not
+        // check valid vouchers twice.
+        // At this point we are done with all recalculations and can tell the voucher list to reset its internal state to be ready
+        // for the next time vouchers are being recalculated.
+        $this->GetActiveVouchers()->allRemoveRunsDone();
```

Comment:
`allRemoveRunsDone` scheint mir als Methodenname problematisch.

Meins du, resetCheckedAndValidState() oder so was?

Reply:
This method is meant to be called by a user of the list and I thought it would make more sense to name it so it is clear WHEN to call it, not why. The user of the class shouldn't be bothered with this in the first place but we cannot get around it at this point. I can rename it but please consider my thoughts. :)

Reply:
I see. Then perhaps 'allRecalculationsCompleted' ? because the caller knows nothing about the remove runs.

Reply:
The recalculations are called like that in the outer abstraction inside the basket. A user of the list class does not call recalculate on the list, it does call RemoveInvalidVouchers.